### ISSUE TITLE
api: add possibility to delete NAMESPACES

### ISSUE THREAD

**Issue by mobuchowski** [MarquezProject/marquez#2095]
Building on top of possibility added in https://github.com/MarquezProject/marquez/pull/2032 - add endpoint to soft delete everything in particular dataset - both jobs and datasets.

**Comment by mobuchowski** [comment]
Few thoughts about deleting namespaces:

Deleting namespace can be thought as a two operations from UI side. 

- All jobs and datasets belonging to the namespace are deleted.
- After delete we can’t see the deleted namespace in Marquez namespace dropdown.

That leaves us with few challenges if we want to have both.

- How would deleted namespace interact with receiving new event for the same namespace?
    - Should the namespace be back in namespace list?
        - This would require event to check and clear potential flag on namespaces.
    - Should we “undelete” the namespace with all the resources?
        - Naive `is_hidden` flag would work exactly like that. Maybe better idea is having all `is_hidden` flags - on jobs, datasets and namespaces set. If event clears namespace flag, it does not clear all flags for jobs and namespaces.
        - Alternative to flags could be logic that shows on the UI only namespaces with some existing jobs or datasets - checking count on every `GET /namespace` call. That is not consistent with `default` namespace though.
    - Should we only show the jobs/datasets that were created by events received after deletion?
        - In this case, should there be any undelete logic for the whole namespace?
- Should behavior of deleted namespace, and namespace from which all the datasets and jobs were deleted be different?
- Theoretically simplest approach, hard delete namespaces would require multiple cascades - everything references namespaces.

**Comment by howardyoo** [comment]
My `two-cents`:
- Should the namespace be back in namespace list? --> Yes
- Should we “undelete” the namespace with all the resources --> Yes
- Should we only show the jobs/datasets that were created by events received after deletion? --> Yes

**Comment by mobuchowski** [comment]
>Should we “undelete” the namespace with all the resources --> Yes
>Should we only show the jobs/datasets that were created by events received after deletion? --> Yes

But those conflict 😉

**Comment by julienledem** [comment]
I agree that the namespace should be marked soft deleted as well. It does not show up in the list of existing namespaces but we can retrieve it.

I think we should adopt the following behavior: Anywhere we would have created a resource (Job, Dataset, Namespace) that didn't exist yet and there was previously an existing but (soft) deleted one we should just undelete it and update it.
Example: 
- when receiving an OL event for a job or dataset that don't exist yet.
- when creating a Dataset or Job in a namespace that doesn't exist yet (from OL event).

In this context a NAMESPACE is either:
- a datasource, the namespace for datasets. This means we deleted a database instance for example and all datasets it contains get (soft) deleted. When new events come in we have to assume it is an empty instance with the same name. If the same datasets exist in it, they will eventually be observed.
- a job scheduler, the namespace for jobs. This means we deleted a scheduler instance and all jobs it runs get soft deleted. If a new scheduler comes in with the same name, the jobs will be observed again.

In both cases, if new events come in with those namespaces that have been deleted previously, we should, behave like we would for a new job or dataset. The namespace gets created automatically (or in this case we just remove the soft delete). 

As a follow up, do we need a parameter on GET that allows seeing soft deletes?
 GET /api/v1/namespaces/foo/datasets?include_deleted=true
